rfd/0201-native-auto-approval.md

```diff
@@ -0,0 +1,364 @@
+---
+authors: Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+
+# RFD 0201 - Native Auto Approval
+
+## Required Approvers
+* Engineering: @r0mant && @fheinecke
+
+## What
+This document describes how Teleport will support native automatic approvals for
+access requests.
+
+## Why
+Currently, Teleport supports automatic approvals of access requests, but support
+is limited. Automatic approvals requires a separate plugin, and only a subset of
+the plugins support automatic approvals. Users requesting this feature do not
+wish to maintain a separate piece of software, or wish to integrate with a
+separate service to utilize this feature. In order to support a wider range of
+use cases, Teleport should support automatic approvals natively.
+
+Automatic approvals enables teams to enforce zero standing privileges, while
+allowing users to get access to their pre approved resources for a limited
+period of time. There are alternative ways to enforce zero standing privileges,
+but a big part of this feature request is to provide a frictionless user
+experience. For example, access lists can be used to achieve similar behavior,
+but a limitation with access lists is that the user must logout and login again
+to renew their certificates to obtain the new roles.
+
+This feature has also been requested for internal use at Teleport. The Tools
+team enforces gated access to certain pipeline activity. The team would like the
+dev environment to closely mimic the production environment and would like to
+enforce the same gated access, but it would reduce a lot of friction if access
+requests can be automatically approved in the dev environment.
+
+## Goals
+1. Native support for automatic approvals. Integration with external service is
+not required.
+2. Automatic approvals can be configured to allow users with certain traits to
+be automatically approved. For example, all `L1` engineers on team `Cloud` in
+location `Seattle` are pre-approved for the access request.
+3. Automatic approvals can be configured to approve access to select resources.
+For example, the requesting user is pre-approved to access all resources in the
+`dev` environment.
+4. Automatic approvals can be configured using Access Monitoring Rules.
```

Comment:
Is this a goal or an implementation detail?

Reply:
This is a UX goal. Previous implementation of automatic approvals relied on annotations within Teleport roles to configure automatic approvals. For example, in the [PagerDuty guide](https://goteleport.com/docs/admin-guides/access-controls/access-request-plugins/ssh-approval-pagerduty/), the `pagerduty_services` annotation is used to configure automatic approval. We would like to deprecate this approach, and rely on AMRs instead for configuring auto approvals.